<H2>Description</H2>
<P>
The
<a target="_blank" href="https://commonfund.nih.gov/GTEx/index">NIH Genotype-Tissue Expression (GTEX) project</a>
was created to establish a sample and data resource for studies on the relationship between 
genetic variation and gene expression in multiple human tissues. 
This track hub shows RNA-seq read coverage signal graphs of GTEx samples in the
GTEx midpoint milestone data release (V6, October 2015).
A total of 7572 tracks representing samples from 570 
adult post-mortem individuals are included.</P>

<H2>Display Conventions</H2>
<P>
The individual sample tracks are organized by gender (Female Donors, Male Donors) and age, 
in 6 tracks, where samples can be selected from a donor vs. tissue matrix.

<ul>
<li>Female 20-49 yrs</li>
<li>Female 50-59 yrs</li>
<li>Female 60-69 yrs</li>
<li>Male 20-49 yrs</li>
<li>Male 50-59 yrs</li>
<li>Male 60-69 yrs</li>
</ul>

An alternative organization is provided by the 'By Tissues' supertrack set, which consists of
a track containing all samples for each of the 53 tissues, with a selection matrix of
donor gender and age.

<H2>Methods</H2>
Tissue samples were obtained using the GTEx standard operating procedures for informed consent
and tissue collection, in conjunction with the 
<a target=_"blank" href="http://biospecimens.cancer.gov/resources/sops/gtex.asp">
National Cancer Institute Biorepositories and Biospecimen</a>.
All tissue specimens were reviewed by pathologists to characterize and
verify organ source.
Images from stained tissue samples can be viewed via the 
<a target="_blank" href="https://brd.nci.nih.gov/brd/image-search/searchhome">
NCI histopathology viewer</a>.
The Qiagen PAXgene non-formalin tissue preservation product was used to stabilize 
tissue specimens without cross-linking biomolecules.</P>
<P>
The Illumina TruSeq protocol was used to create an unstranded polyA+ library sequenced
on the Illumina HiSeq 2000 platform to produce 76-bp paired end reads at a depth 
averaging 50M aligned reads per sample.
<P>
Sequence reads for this track were aligned to the hg38/GRCh38 human genome using STAR2
assisted by the GENCODE v24 transcriptome definition. The read coverage signal graphs
were produced using the --outWig option of STAR2.  Read mapping was performed at UCSC by the 
Computational Genomics lab, using the Toil pipeline. The resulting bedGraph
files were converted to bigWig format for display.  The bedGraph files were coordinate 
converted to the hg19/GRCh37 assembly via the UCSC liftOver tool.
<P>

<H2>Subject and Sample Characteristics</H2>
<P>
The scientific goal of the GTEx project required that the donors and their biospecimen 
present with no evidence of disease. 
The tissue types collected were chosen based on their clinical significance, logistical 
feasibility and their relevance to the scientific goal of the project and the 
research community. 
Postmortem samples were collected from non-diseased donors with ages ranging from 20 to 79. 34.4% of donors were female and 65.6% male. 
<div> <img border=1 src='/images/gtex/gtexSampleRin.V6.png'></div>
<p></p>
<div><img border=1 src='/images/gtex/gtexSampleAge.V6.png'></div></p>
<p>
Additional summary plots of GTEx sample characteristics are available at the 
<a target="_blank" href="http://gtexportal.org/home/tissueSummaryPage">
GTEx Portal Tissue Summary</a> page.</p>

<H2>Credits</H2>
<P>
Samples were collected by the GTEx Consortium.
RNA-seq was performed by the GTEx Laboratory, Data Analysis and Coordinating Center 
(LDACC) at the Broad Institute.
John Vivian, Melissa Cline, and Benedict Paten of the UCSC Computational Genomics lab were
responsible for the sequence read mapping and signal file generation used to produce this hub.
Kate Rosenbloom and Parisa Nejad of the UCSC Genome Browser group were responsible for 
data file post-processing and track hub configuration.
</p>

<H2>Contacts</H2>
<P>
For questions about the GTEx data, contact the 
<A HREF="mailto:&#103;&#116;&#101;&#120;-&#104;&#101;&#108;p&#64;&#98;&#114;&#111;&#97;&#100;in&#115;&#116;&#105;&#116;&#117;t&#101;.
&#111;&#114;&#103;">
GTEx LDACC at the Broad Institute</A>.
<!-- above address is gtex-help at broadinstitute.org -->
For questions about this track hub, contact the 
<A HREF="http://genome.ucsc.edu/contacts.html">
UCSC Genome Browser mailing list</A>.</P>

<H2>References</H2>
<p>
GTEx Consortium.
<a href="http://dx.doi.org/10.1038/ng.2653" target="_blank">
The Genotype-Tissue Expression (GTEx) project</a>.
<em>Nat Genet</em>. 2013 Jun;45(6):580-5.
PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23715323" target="_blank">23715323</a>; 
PMC: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4010069/" target="_blank">PMC4010069</a> </p>
<!--
<p>
Vivian J, Rao A, Nothaft F, Ketchum C, Armstrong J, Novak A, Narkizian J, Deran AD, Musselman-Brown A, Schmidt H, Amstutz P, Frazer S, Craft B, Goldman M, Rosenbloom KR, Cline M, O'Connor B, Hanna M, Birger C, Kent WJ, Patterson P, Zhu J, Zaranek S, Getz G, Haussler D, Paten P.
Rapid and efficient analysis of 20,000 RNA-seq samples with Toil.
<em>Nature Biotech</em>. Submitted 2016 May.
-->
<p>
Carithers LJ, Ardlie K, Barcus M, Branton PA, Britton A, Buia SA, Compton CC, DeLuca DS, Peter-Demchok J, Gelfand ET <em>et al</em>.
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/pmid/26484571/" target="_blank">
A Novel Approach to High-Quality Postmortem Tissue Procurement: The GTEx Project</a>.
<em>Biopreserv Biobank</em>. 2015 Oct;13(5):311-9.
PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26484571" target="_blank">26484571</a>; 
PMC: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4675181/" target="_blank">PMC4675181</a></p>

Mel&#233; M, Ferreira PG, Reverter F, DeLuca DS, Monlong J, Sammeth M, Young TR, Goldmann JM,
Pervouchine DD, Sullivan TJ <em>et al</em>.
<a href="http://www.sciencemag.org/cgi/pmidlookup?view=long&amp;pmid=25954002" target="_blank">
Human genomics. The human transcriptome across tissues and individuals</a>.
<em>Science</em>. 2015 May 8;348(6235):660-5.
PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25954002" target="_blank">25954002</a>; PMC: <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4547472/" target="_blank">PMC4547472</a></p>

<p>
DeLuca DS, Levin JZ, Sivachenko A, Fennell T, Nazaire MD, Williams C, Reich M, Winckler W, Getz G.
<a href="http://bioinformatics.oxfordjournals.org/cgi/pmidlookup?view=long&amp;pmid=22539670"
target="_blank">
RNA-SeQC: RNA-seq metrics for quality control and process optimization</a>.
<em>Bioinformatics</em>. 2012 Jun 1;28(11):1530-2.
PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22539670" target="_blank">22539670</a>; PMC: <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3356847/" target="_blank">PMC3356847</a></p>

