<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2//EN">

<HTML>
  <HEAD>
    <META name="generator" content=
    "HTML Tidy for Linux/x86 (vers 1st June 2002), see www.w3.org">

    <TITLE>GenBank/RefSeq Alignment Step</TITLE>
  </HEAD>

  <BODY bgcolor="white">
    <H1>GenBank/RefSeq Alignment Step</H1>
    This step is done for each browser database (species and assembly) that is
    being updated. 

    <H3>Algorithm</H3>

    <UL>
      <LI>Select the most current GenBank (and corresponding RefSeq) full
      releases from the <CODE>processed/</CODE> directory.</LI>

      <LI>
        If there is not a corresponding <CODE>aligned/<EM>...</EM>/full</CODE>
        directory for the release: 

        <UL>
          <LI>If there is a previous aligned release for this database, find
          the aligned sequences that have not changed since that release and
          the new, full release. Save these in a a temporary directory.</LI>

          <LI>Copy new and changed sequences to temporary fasta files for
          alignment.</LI>
        </UL>
      </LI>

      <LI>
        For each <CODE>processed/<EM>...</EM>/daily.<EM>${ver}</EM>/</CODE>
        directory that does not have a completed <CODE>alignment/</CODE>
        directory: 

        <UL>
          <LI>Copy new and changed sequences to temporary fasta files for
          alignment.</LI>
        </UL>
      </LI>

      <LI>If the number of sequences requiring alignment exceeds some
      configured threshold, send e-mail requesting an alignment on the big
      cluster and stop the automated process. Normally, this should only be
      required when a new database is built.</LI>

      <LI>If the number of sequences is below the threshold, run BLAT on the
      mini-cluster, using the <CODE>parasol make</CODE> facility.</LI>

      <LI>
        Process the completed alignments: 

        <UL>
          <LI>Combine alignments migrated from the previous releases if
          pending</LI>

          <LI>Building an index file</LI>

          <LI>Checksum the files.</LI>

          <LI>Do some sanity check on the alignment. Check that changed
          sequences continue to align, at least in most cases and that the
          number of aligned sequences increases.</LI>
        </UL>
      </LI>
    </UL>

    <H3><A name="dir">Directory structure</A></H3>

    <UL>
      <LI>
        <CODE>$gbRoot/data/aligned/</CODE> - aligned files 

        <UL>
          <LI>
            <CODE>genbank.<EM>${ver}</EM>/</CODE> 

            <UL>
              <LI>
                <CODE><EM>${db}</EM>/</CODE> - alignments for this genome
                database (e.g. hg12). 

                <UL>
                  <LI>
                    <CODE>full/</CODE> - Alignments corrisponding to the full
                    release. This is a combination alignments migrated from
                    previous releases and new alignments. 

                    <UL>
                      <LI><CODE>mrna.native.psl.gz</CODE>,
                      <CODE>mrna.native.oi.gz</CODE>,
                      <CODE>mrna.native.alidx</CODE>,
                      <CODE>mrna.native.md5</CODE></LI>

                      <LI><CODE>est.aa.native.psl.gz</CODE>,
                      <CODE>est.aa.native.oi.gz</CODE>,
                      <CODE>est.aa.native.alidx</CODE>,
                      <CODE>est.aa.native.md5</CODE>,</LI>

                      <LI><CODE>mrna.xeno.psl.gz</CODE>,
                      <CODE>mrna.xeno.oi.gz</CODE>,
                      <CODE>mrna.xeno.alidx</CODE>,
                      <CODE>mrna.xeno.md5</CODE></LI>

                      <LI><CODE>est.aa.xeno.psl.gz</CODE>,
                      <CODE>est.aa.xeno.oi.gz</CODE>,
                      <CODE>est.aa.xeno.alidx</CODE>,
                      <CODE>est.aa.xeno.md5</CODE>,</LI>
                    </UL>
                  </LI>

                  <LI><CODE>daily.<EM>${date}</EM>/</CODE> - Alignments for
                  sequence that were new or modified in the daily update.</LI>
                </UL>
              </LI>
            </UL>
          </LI>

          <LI>
            <CODE>refseq.<EM>${ver}</EM>/</CODE> - BLAT alignments for RefSeq,
            same structure as used for GenBank, with only native mRNAs. 

            <UL>
              <LI>...</LI>
            </UL>
          </LI>
        </UL>
      </LI>
    </UL>

    <H3>Index file</H3>
    Two alignment index files are always created for the corrisponding
    processed.gbidx file, for native and xeno, the if there are no sequences to
    align. This supports easy checking for the alignment being completed. The
    file is a tab-seperated in the format: 
<PRE>
    acc version numaligns
</PRE>
    The name of the file is either <CODE>mrna.alidx</CODE> or
    <CODE>est.*.alidx</CODE> and is associated with the a <CODE>*.psl</CODE>
    file of the same name. The columns are: 

    <UL>
      <LI><CODE>acc</CODE> - GenBank or RefSeq accession</LI>

      <LI><CODE>version</CODE> - Version number, not including the
      accession</LI>

      <LI><CODE>numAligns</CODE> - Count of the number of alignments for this
      accession.</LI>
    </UL>
  </BODY>
</HTML>

