<!DOCTYPE html PUBLIC "-//IETF//DTD HTML 2.0//EN">

<HTML>
  <HEAD>
    <TITLE>Genbank configuration file</TITLE>
  </HEAD>

  <BODY bgcolor="white">
    <H1>Genbank configuration file</H1>
    The file <CODE><EM>$gbRoot</EM>/etc/genbank.conf</CODE>
    contains options control the alignment and loading of
    GenBank and RefSeq data.  The file is in the same format as
    <CODE>.hg.conf</CODE>:
    <UL>
    <LI> blank lines or lines where the first non-blank character is a &quot;#&quot; are ignored.
    <LI> records are lines of:
         <blockquote>
         <CODE><EM>name</EM>=<EM>value</EM></CODE>
         </blockquote>
    <LI> By convention, names consist of dot separated words used to define a hierarchy.  Yes, this
         is ugly and verbose.
    </UL>
    <H3>Variables</H3>
    Variables are available in the configuration file.  A variable is defined using the syntax:
    <blockquote>
         <CODE>var.<EM>varname</EM>=<EM>value</EM></CODE>
    </blockquote>
    Variables maybe referenced in any value using the syntax:
    <blockquote>
         <CODE>${<EM>varname</EM>}</CODE>
    </blockquote>
    A variable must be defined before it is referenced in the file, and are expanded immediately.
    A variable definition may reference another variable.
    <p>
    The program <code>gbConf</code> can be used to print the configuration file with variables
    expanded for debugging purposes.

    <H3>Global configuration parameters</H3>
    <UL>
    <LI> <CODE>cluster.rootDir</CODE> - root directory on cluster filesystems.
    <LI> <CODE>cluster.paraHub</CODE> - Host where parasol hub is running.
    <LI> <CODE>grepIndex.genbank.default</CODE> - Create <EM>grepIndex</EM> files
         for databases in the specified directory.  This value should
         correspond to the value <CODE>grepIndex.genbank</CODE> in the
         browser <CODE>hg.conf</CODE> file.  The <CODE>default</CODE> in this name
         leaves open the possibility of per-server overrides of this
         value, however this is not currently implemented.
    <li> <code>gbdb.genbank</code> - Location of genbank <CODE>gbdb</CODE>
         directory.  Defaults to <CODE>/gbdb/genbank</CODE>.
    <li> <code>build.server</code> - Build server host name.  If set,
         build scripts will generate an error if not run on this server.
    </UL>
    <H3>per-database configuration parameters</H3>
    Per-database configuration parameters start with the database name,
    represented here by <CODE><EM>$db</EM></CODE>.  Default values are set with a
    database name of <CODE>default</CODE>

    <UL>
     <li> <code><em>$db</em>.align.window</code> - Size of genome alignment windows, in bases.
          The genome is partitioned in segments of no more than this sizes for alignment.
     <li> <code><em>$db</em>.align.overlap</code> - Number of bases of overlap in the alignment
          windows.
     <li> <code><em>$db</em>.align.maxGap</code> - Gaps no larger than this
          value are contained within a window rather then starting a 
          new window.  This allows gaps within introns.
     <li> <code><em>$db</em>.align.maxJobTarget</code> - Approximate limit
          on the number of target window bases in a job.  This trades
          of parallelism for the cost of create lots of small jobs
          and their associated files.
     <li> <code><em>$db</em>.align.unplacedChroms</code> - 
          White-space separated list of pseudo-chromosome that contain
          unplaced sequences.  Alignments will not be allowed to span
          gaps on these chromsomes. Requires a lift file to specify the
          gaps.  The names can also be patterns similar to file name
          glob patterns that are matched against chromosome names.  The
          entire pattern must be matched. The meta characters are <CODE>*</CODE>
          and <CODE>?</CODE>.  For example <CODE>*_random</CODE>.
     <li> <code><em>$db</em>.align.minUnplacedSize</code> - Skip
          unplaced sequences smaller than this size.
     <li> <code><em>$db</em>.align.querySplitSize</code> - Max size of query
       FASTA files to create, in megabytes.
     <li> <code><em>$db</em>.serverGenome</code> - glob pattern or file for genome
          sequences on server. This can be NIBs or a TwoBit file.
     <li> <code><em>$db</em>.clusterGenome</code> - glob pattern or file for genome
          sequences on cluster. This can be NIBs or a TwoBit file. The
          resulting list of files must match was what is on the server.
     <li> <code><em>$db</em>.ooc</code> - path to blat ooc file on cluster, or 
          <code>no</code> if no blat ooc is to be used.
     <li> <code><em>$db</em>.maxIntron</code> - blat -maxIntron value.
     <li> <code><em>$db</em>.lift</code> - path to lift file for genome,
          or <CODE>no</CODE> if none.  Lift file is used to find gaps when partitioning the
          sequence for alignment.  It does not have to contain all chromosomes.
          Sequences named <code>gap</code> are treated as gaps as well as gaps in
          contig placement.  A lift file generate by <code>gapToLift</code> is ideal.
     <li> <code><em>$db</em>.hapRegions</code> - path to a PSL that contains
          alignments of haplotype pseudo-chromosomes to the reference chromosomes
          or <CODE>no</CODE> if none.  This is used to map
          alignments between haplotype regions for the near-best in genome alignment
          mappings.  This is not access from cluster jobs.
     <li> <code><em>$db</em>.<em>$srcDb</em>.<em>$cdnaType</em>.<em>$orgCat</em>.pslCDnaFilter</code> - Arguments
           for <CODE>pslCDnaFilter</CODE> for various types of alignments
           or <CODE>no</CODE> to skip pslCDnaFilter.
           Special handling is done for the the <CODE>-polyASizes</CODE> option.
           This option should be supplied without an argument.  The location
           of the generated file will be added when <code>pslCDnaFilter</code>
           is called.
           
          <p>The following values are recongized in the name:
          <ul>
          <li> <code><em>$srcDb</em></code> - <code>genbank</code>, <code>refseq</code>
          <li> <code><em>$cdnaType</em></code> - <code>mrna</code>, <code>est</code>
          <li> <code><em>$orgCat</em></code> - <code>native</code>, <code>xeno</code>
          </ul>

     <li> <code><em>$db</em>.mgcTables.<em>$host</em></code> - indicates which
          MGC tables to create when loading database on <CODE><em>$host</em></code>.
          Values are <CODE>no</CODE>, <CODE>all</CODE>, or <CODE>full</CODE>.
          <CODE><em>$host</em></code> is the value returned by <CODE>uname -n</CODE>.
     <li> <code><em>$db</em>.mgcTables.default</code> - indicates which
          MGC tables to create when there is no host-specific setting.

     <li> <code><em>$db</em>.upstreamGeneTbl</code> - If specified, use this
       table to create upstream FASTA files. It is an error if the table
       doesn't exist.  If not specified, don't create upstream
       FASTAs.

     <li> <code><em>$db</em>.upstreamMaf</code> - If specified, create update
       MAFs using these MAF tables and organism lists, and the genes
       in <code><em>$db</em>.upstreamGeneTbl</code>.  The value should be
       white-space separated pairs of table name and fully qualified path
       to file with order list of organisms to include in the upstream MAF.

     <li> <code><em>$db</em>.ccds.ncbiBuild</code> - includes NCBI build number
       (e.g. 36.3) for CCDS table building.  Specifying this enabled auto-update
       of ccdsGene and related tables.

     <li> <code><em>$db</em>.mgc</code> - Should MGC tables be loaded.
          Values are <CODE>yes</CODE>, or <CODE>no</CODE>.
     <li> <code><em>$db</em>.orfeome</code> - Should ORFeome tables be loaded.
          Values are <CODE>yes</CODE>, or <CODE>no</CODE>.

     <li> <code><em>$db</em>.perChromTables</code> - Set to <CODE>no</CODE>
          if per-chromosome alignment tables should not be created.

     <li> <code><em>$db</em>.<em>$srcDb</em>.<em>$cdnaType</em>.<em>$orgCat</em>.load</code> -
          Should cDNAs be aligned and loaded into the database for the
          specified category. Value is <CODE>yes</CODE> or <CODE>no</CODE>.
     <li> <code><em>$db</em>.<em>$srcDb</em>.<em>$cdnaType</em>.<em>$orgCat</em>.align</code> -
          Should cDNAs be aligned for the
          specified category. Value is <CODE>yes</CODE> or <CODE>no</CODE>.  This is used if it's
          desirable to align a particular category but not create a track of it.
     <li> <code><em>$db</em>.<em>$srcDb</em>.<em>$cdnaType</em>.<em>$orgCat</em>.loadDesc</code>
          Should the description table  be loaded? Value is <CODE>yes</CODE> or <CODE>no</CODE>.
          Descriptions are normally not loaded for ESTs as they are
          large and not very useful
     <li> <code><em>$db</em>.downloadDir</code> - directory relative to <CODE>goldenPath/</CODE>
           where download files are stored.
     <li> <code><em>$db</em>.align.prefilter</code> - Prefilter alignments in cluster jobs.
           If set to <CODE>yes</CODE> or not specified do prefiltering. If set to <CODE>no</CODE>,
           no prefiltering is done.  This is often useful for debugging.

     <li> <code><em>$db</em>.<em>$srcDb</em>.mrna.blatTargetDb</code> -
          Should BLAT targetDb two-bit files be created for database for the
          specified srcDb. Value is <CODE>yes</CODE> or <CODE>no</CODE>.
    </UL>
  </BODY>
</HTML>
